File: 2022/day_4/main.py
def main() -> None:
    with open('input.txt') as f:
        pairs: list[str] = f.read().splitlines()
    
    subset_counter: int = 0
    overlap_counter: int = 0
    for pair in pairs:
        elf_1, elf_2 = pair.split(',')
        elf_1_set: set[int] = set([int(x) for x in range(int(elf_1.split('-')[0]), int(elf_1.split('-')[1]) + 1)])
        elf_2_set: set[int] = set([int(x) for x in range(int(elf_2.split('-')[0]), int(elf_2.split('-')[1]) + 1)])

        if elf_1_set.issubset(elf_2_set) or elf_2_set.issubset(elf_1_set):
            subset_counter += 1

        if elf_1_set.intersection(elf_2_set):
            overlap_counter += 1
            

    print(f'Part 1. Number of subsets: {subset_counter}')
    print(f'Part 2. Number of overlaps: {overlap_counter}')



if __name__ == "__main__":
    main()

File: 2022/day_4/main.go
package main

import (
    "fmt"
    "io/ioutil"
    "strings"
    "strconv"
)

func main() {
    content, error := ioutil.ReadFile("input.txt")
    if error != nil {
        fmt.Println("file not found")
        panic(error)
    }

    subsetCounter := 0
    overlapCounter := 0
    for _, pair := range strings.Split(string(content), "\n") {
        if pair == "" {
            continue
        }
        elfPair := strings.Split(pair, ",")
        
        assignment1 := strings.Split(elfPair[0], "-")
        assignment2 := strings.Split(elfPair[1], "-")

        a1S, _ := strconv.Atoi(assignment1[0])
        a1E, _ := strconv.Atoi(assignment1[1])
        a2S, _ := strconv.Atoi(assignment2[0])
        a2E, _ := strconv.Atoi(assignment2[1])

        if (a1S <= a2S && a1E >= a2E) || (a1S >= a2S && a1E <= a2E) {
            subsetCounter++
        }
        if (a1S <= a2S && a1E >= a2S) || (a1S >= a2S && a1S <= a2E) {
            overlapCounter++
        }
    }

    fmt.Println("Number of subsets:", subsetCounter)
    fmt.Println("Number of overlaps:", overlapCounter)
}
